### PR TITLE
[sinks] Fail open when we can't fetch Kafka config

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

This PR fixes a previously unreported bug.

Kafka allows specifying the partition number and replication factor when creating a topic. When -1 is specified, it will use a broker-wide default value. However, older versions of Kafka don't have the -1 defaulting behaviour... as a workaround, we fetch the config and apply the defaults manually.

If fetching the config fails, we fail the entire ensure-creation request, even though it is actually fairly likely to succeed: either because the topic exists, or because the Kafka version is recent enough to handle these defaulted configs. This PR "fails open", replacing the error with a warning log and continuing on.

### Tips for reviewer

This is mostly code movement; you may want to hide whitespace changes with `?w=1`.

I'll try and reproduce this locally before merging.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by benesch — APPROVED** [review]
Code change LGTM. Do you think it's possible to write a test for this?

**Comment by bkirwi** [comment]
Thanks for the look!

> Do you think it's possible to write a test for this?

Theoretically! Though I'll be more confident once I've been able to reproduce this locally.

**Comment by bkirwi** [comment]
Okay, I have reproduced this locally.
- I could not reproduce this by leaving the config unset in the broker configuration. Even when the explicit config is removed from `server.properties`, it appears to fall back to a value of `1`.
- However! I did manage to produce this error by blocking `DescribeConfigs` on the `cluster` resource via ACLs. (I think the fact that we're getting an unhelpful error may be a rdkafka bug, but this looks like it would fail either way.) This seems to be the only place where we need this ACL, at least in the sink; with the patch on this branch, the sink is able to start and make progress.

Writing an automated test for this will be annoying and I don't think I can get it out today, though I should be able to follow up on it later.

**Comment by benesch** [comment]
> Writing an automated test for this will be annoying and I don't think I can get it out today, though I should be able to follow up on it later.

@philip-stoev or someone else from the QA team can probably help with this!

**Comment by philip-stoev** [comment]
@bkirwi can you share how you blocked `DescribeConfigs` when you ran things manually?

**Comment by benesch** [comment]
@philip-stoev here's a good guide on the subject: https://jaceklaskowski.gitbooks.io/apache-kafka/content/kafka-demo-acl-authorization.html

tl;dr is set `authorizer.class.name=kafka.security.authorizer.AclAuthorizer` in the properties, and then add an ACL that allows writing to a topic but _not_ describing configs. I think something like this would do the trick:

```
kafka-acls.sh \
  --bootstrap-server :9093 \
  --add \
  --allow-principal User:CN=whoever \
  --operation Write \
  --topic '*' \
  --command-config /tmp/kafka-ssl-demo/root.properties
```

Probably want to adapt one of the existing kafka-auth tests, since ACLs only work in tandem with authentication.

**Comment by philip-stoev** [comment]
I gave a Deny ACL to "ALL" against all topics for a user, and yet `CREATE SINK` fails all the way at:

```
aterialize=> CREATE SINK no_describe_config_sink FROM no_describe_config_view INTO KAFKA CONNECTION no_describe_config_kafka_conn (TOPIC 'testdrive-no-describe-config-XXXXX', replication factor = -1 , partition count = -1 ) FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION no_describe_config_csr_conn ENVELOPE DEBEZIUM;
ERROR:  error registering kafka topic for sink: Error creating topic testdrive-no-describe-config-XXXXX for sink: unable to fetch topic metadata after creation
```

which I think is too late. Please stand by.

**Comment by philip-stoev** [comment]
Ok so I was able to create a Kafka user that is not allowed to run `DescribeConfigs` against the `cluster` resource:

```
kafka-acls \
  --bootstrap-server kafka:9092 \
  --add \
  --deny-principal User:CN=no_describe_config \
  --operation DescribeConfigs \
  --cluster \
  --command-config //tmp/foo.properties
```

However, this does not cause the expected error code path to trigger. In the `discover_topic_configs` method, I get the following:

For the `configs` variable:

```
[Ok(ConfigResource { specifier: Broker(1), entries: [] })]
```

For the `config` variable:

```
ConfigResource { specifier: Broker(1), entries: [] }
```

In other words, an empty list of `entries` is returned by `client.describe_configs` , rather than an `Err()` or something else.

that is, neither of the error paths in the method will trigger. We get to the `for entry in config.entries` loop, match nothing with it, and therefore the method returns the `-1` defaults for both settings defined at the top of the method.

Is that expected -- I was under the impression that a failure is expected.

In fact, from a cursory examination of the code, rdkafka's `describe_configs` method is unlikely to return an `Err`, there is only one `?` operator in the entire thing.

**Comment by bkirwi** [comment]
Yep, that sounds like the error condition I was seeing! My hypothesis is that the broker is passing along an error code that rdkafka is ignoring. (When I make the equivalent call from a JVM client, I get an exception.) But it seems clear that either way this array is empty.

> However, this does not cause the expected warning path to trigger.

Is the surprise that the log entry is not showing up? [I removed those errors](https://github.com/MaterializeInc/materialize/pull/18853/files?diff=unified&w=1#diff-3f5cfda6fea4a85143ff18d5c4b11af731b1c0819257865a6a22d06691f48ef6L123) since we're failing open. But I can put them back if that would be more clear!

In either case, it sounds like your test would fail to create the sink on `main` but pass on this branch? If so that definitely sounds worth merging in.

**Comment by bkirwi** [comment]
> But I can put them back if that would be more clear!

I pushed a change to error when the resulting config list for our cluster is empty, since it seems unlikely that this will ever happen in a non-error case. Let me know if this is what you had in mind!

**Review by philip-stoev — APPROVED** [review]
I pushed a test that passes with this patch but fails on main.

**Comment by aljoscha** [comment]
tyvm, @philip-stoev! I was hoping you already had a test up your sleeve :sweat_smile:

**Comment by philip-stoev** [comment]
I did not have a test, but I have gained something -- a deep, simmering hatred for Kafka authentication.

File: test/kafka-sasl-plain/mzcompose.py [hunk 1]
```diff
@@ -118,4 +123,22 @@ def workflow_default(c: Composition) -> None:
     c.up("test-certs")
     c.up("zookeeper", "kafka", "schema-registry")
     c.up("materialized")
-    c.run("testdrive", "*.td")
+
+    c.run("testdrive", "smoketest.td")
+
+    # Deny the DescribeConfigs cluster priviilege to user no_describe_configs
+    c.exec(
```
Comment (aljoscha) [inline test/kafka-sasl-plain/mzcompose.py:130]: typo: priviilege

**Review by aljoscha — APPROVED** [review]

File: test/kafka-sasl-plain/mzcompose.py [hunk 2]
```diff
@@ -118,4 +123,22 @@ def workflow_default(c: Composition) -> None:
     c.up("test-certs")
     c.up("zookeeper", "kafka", "schema-registry")
     c.up("materialized")
-    c.run("testdrive", "*.td")
+
+    c.run("testdrive", "smoketest.td")
+
+    # Deny the DescribeConfigs cluster priviilege to user no_describe_configs
```
Comment (aljoscha) [inline test/kafka-sasl-plain/mzcompose.py:129]:
```suggestion
    # Deny the DescribeConfigs cluster privilege to user no_describe_configs
```

**Review by petrosagg — APPROVED** [review]
